File: Data Structures & Algorithms/count-number-of-islands/submission-3.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]: return 0
        ans = 0

        def dfs(row, col):
            dirs = [[0,1], [1,0], [-1,0], [0,-1]]

            if (row < 0 or row >= ROWS or col < 0 or col >= COLS or grid[row][col] == '0'): return

            grid[row][col] = '0'
            for nr, nc in dirs:
                dfs(row + nr, col + nc)

        ROWS, COLS = len(grid), len(grid[0])
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == '1':
                    ans += 1
                    dfs(r, c)
        return ans

File: Data Structures & Algorithms/max-area-of-island/submission-3.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        if (grid.empty()) return 0;

        int res = 0;
        int ROWS = grid.size(), COLS = grid[0].size();

        for (int r = 0; r < ROWS; r++){
            for (int c = 0; c < COLS; c++){
                if (grid[r][c] == 1) {
                    int tmp = dfs(r, c, grid, ROWS, COLS);
                    res = max(res, tmp);
                }
            }
        }
        return res;
    }

    int dfs(int row, int col, vector<vector<int>>& grid, int ROWS, int COLS){
        if (row < 0 || col < 0 || row >= ROWS || col >= COLS || grid[row][col] == 0){
            return 0;
        }

        grid[row][col] = 0;
        int area = 1;
        vector<pair<int, int>>dirs = {{1,0}, {0,1}, {-1,0}, {0,-1}};

        for (auto& [dr, dc] : dirs){
            int nr = row + dr, nc = col + dc;
            area += dfs(nr, nc, grid, ROWS, COLS);
        }
        return area;
    }
};